server/src/main/java/org/opensearch/search/sort/ScriptSortBuilder.java

```diff
@@ -403,29 +408,32 @@ public BucketedSort newBucketedSort(
                 final NumberSortScript.Factory numberSortFactory = context.compile(script, NumberSortScript.CONTEXT);
                 final NumberSortScript.LeafFactory numberSortScript = numberSortFactory.newFactory(script.getParams(), context.lookup());
                 return new DoubleValuesComparatorSource(null, Double.MAX_VALUE, valueMode, nested) {
-                    NumberSortScript leafScript;
+                    // introducing a map to keep a mapping between the leaf reader context and leaf script
+                    // such that the functions of the class are thread safe in case of concurrent search
+                    final Map<LeafReaderContext, NumberSortScript> leafContextSortScriptMap = new ConcurrentHashMap<>();
 
                     @Override
                     protected SortedNumericDoubleValues getValues(LeafReaderContext context) throws IOException {
-                        leafScript = numberSortScript.newInstance(context);
+                        NumberSortScript leafScript = numberSortScript.newInstance(context);
```

Comment:
similar change to use `final` and use that for other methods

Reply:
done